cs/azure/Trainer/LearnEventProcessorFactory.cs

```diff
@@ -0,0 +1,247 @@
+                    }
+
+                    performanceCounters.Stage1_JSON_DeserializePerSec.Increment();
+
+                    // delayed
+                    if (data.Example == null)
+                    {
+                        this.performanceCounters.Feature_Requests_Pending.Increment();
+                        yield break;
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                this.telemetry.TrackException(ex, new Dictionary<string, string> { { "JSON", data.JSON } });
+
+                this.performanceCounters.Stage2_Faulty_Examples_Total.Increment();
+                this.performanceCounters.Stage2_Faulty_ExamplesPerSec.Increment();
+
+                yield break;
+            }
+
+            yield return data;
+        }
+
+        private IEnumerable<object> ProcessEvent(object evt)
+        {
+            // single threaded loop
+            var eventHubExample = evt as PipelineData;
+            if (eventHubExample != null)
+            {
+                yield return this.trainer.Learn(eventHubExample);
+
+                // this.latencyOperation.Process(result);
+
+                if (this.trainer.ShouldCheckpoint())
+                {
+                    var checkpointData = this.trainer.CreateCheckpointData();
+                    checkpointData.UpdateClientModel = true;
+                    yield return checkpointData;
+                }
+            }
+            else if (evt is CheckpointEvent)
+                yield return this.trainer.CreateCheckpointData();
+            else
+                this.telemetry.TrackTrace($"Unsupported stage 2 event '{evt}'", SeverityLevel.Warning);
+        }
+
+        public IEventProcessor CreateEventProcessor(PartitionContext context)
+        {
+            return new LearnEventProcessor(this, this.performanceCounters);
+        }
+
+        public ITargetBlock<PipelineData> DeserializeBlock { get { return this.deserializeBlock; } }
+
+        public ITargetBlock<object> LearnBlock { get { return this.learnBlock; } }
+
+        /// <summary>
+        /// Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.
+        /// </summary>
+        public void Dispose()
+        {
+            if (this.learnBlock != null)
+            {
+                // complete beginning of the pipeline
+                this.deserializeBlock.Complete();
+
+                // wait at the end of the pipeline
+                this.checkpointBlock.Completion.Wait(TimeSpan.FromMinutes(1));
+            }
+
+            if (this.evalOperation != null)
+            {
+                this.evalOperation.Dispose();
+                this.evalOperation = null;
+            }
+
+            //if (this.latencyOperation != null)
+            //{
+            //    this.latencyOperation.Dispose();
+            //    this.latencyOperation = null;
+            //}
```

Comment:
I actually want to implement this in the future... just not enough time now

---
In reply to: [66973362](https://github.com/JohnLangford/vowpal_wabbit/pull/1013#discussion_r66973362) [](ancestors = 66973362)